src/Parser.ts

```
@@ -0,0 +1,122 @@
+          return;
+        }
+
+        const fileName = utils.parseFileName(view);
+        const fileSize = utils.extractOctal(
+          view,
+          HeaderOffset.FILE_SIZE,
+          HeaderSize.FILE_SIZE,
+        );
+        const fileMtime = new Date(
+          utils.extractOctal(
+            view,
+            HeaderOffset.FILE_MTIME,
+            HeaderSize.FILE_MTIME,
+          ),
+        );
+        const fileMode = utils.extractOctal(
+          view,
+          HeaderOffset.FILE_MODE,
+          HeaderSize.FILE_MODE,
+        );
+        const ownerGid = utils.extractOctal(
+          view,
+          HeaderOffset.OWNER_GID,
+          HeaderSize.OWNER_GID,
+        );
+        const ownerUid = utils.extractOctal(
+          view,
+          HeaderOffset.OWNER_UID,
+          HeaderSize.OWNER_UID,
+        );
+        const ownerName = utils.extractChars(
+          view,
+          HeaderOffset.OWNER_NAME,
+          HeaderSize.OWNER_NAME,
+        );
+        const ownerGroupName = utils.extractChars(
+          view,
+          HeaderOffset.OWNER_GROUPNAME,
+          HeaderSize.OWNER_GROUPNAME,
+        );
+        const ownerUserName = utils.extractChars(
+          view,
+          HeaderOffset.OWNER_USERNAME,
+          HeaderSize.OWNER_USERNAME,
+        );
+        const fileType =
+          utils.extractChars(
+            view,
+            HeaderOffset.TYPE_FLAG,
+            HeaderSize.TYPE_FLAG,
+          ) === EntryType.FILE
+            ? 'file'
+            : 'directory';
+
+        if (fileType === 'file') {
+          this.state = 'header';
+          this.remainingBytes = fileSize;
+        }
+
+        const parsedHeader: Header = {
+          type: 'header',
+          fileType,
+          fileName,
+          fileMode,
+          fileMtime,
+          fileSize,
+          ownerGid,
+          ownerUid,
+          ownerName,
+          ownerUserName,
+          ownerGroupName,
+        };
+
+        return parsedHeader;
```

The parser logic needs to be separated out of the state machine as parser functions.

So we need the base level parser functions like 
- `ParseHeader`.
- `ParseData`.
- `ParseNull`.

These will validate that the data fits the expected structure, but also returns the expected token for us. So if the state machine is waiting for a header chunk and we get not that, we need to throw a parsing error. Data is pretty freeform so nothing to validate there. We also need have parsing errors in the case where we get one null header but then get something else.

You also need to account for the data ending early, so we need an `end` to signify the end of the stream. if this is called before the end of the archive then we need to throw a parser error.

---

Currently I have the errors simply called `ErrorVirtualTar`, but I will probably segregate them further for parsers and generators. As it will become too long with the full name, I will concatenate the name to `ErrorTar` for brevity.

---

No `ErrorVirtualTar` sounds good.